### PR TITLE
🗃️ feat(database): add resource_permissions table and messenger credential columns

### PR DESCRIPTION
**Pull request by rdmclin2**
#### 💻 Change Type

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 👷 build
- [ ] ⚡️ perf
- [ ] ✅ test
- [ ] 📝 docs
- [ ] 🔨 chore

#### 🔀 Description of Change

DB-only PR consolidating two schema changes into a single migration `0123_resource_permissions_and_messenger_credentials`; model/service/UI usage lands in follow-up PRs.

**1. New `resource_permissions` table** for workspace-level resource access control (General Access model):

- Each row grants a workspace-scoped access level (`access_level`) on a resource identified by `resource_type` + `resource_id`; absence of a row falls back to the default access semantics.
- Unique index on `(workspace_id, resource_type, resource_id)` so each resource has at most one access-level entry per workspace.
- Secondary indexes on `(resource_type, resource_id)` and `(workspace_id)` for lookup paths.
- FKs: `workspace_id → workspaces.id` (cascade delete), `created_by → users.id` (set null).

**2. Two nullable columns on `messenger_account_links`** for user-scoped credential platforms (e.g. WeChat):

- `application_id` (`varchar(255)`): platform-side application/bot id, used to fail closed on inbound payloads before decrypting credentials. Null for shared-bot platforms.
- `credentials` (`text`): AES-GCM encrypted platform credential JSON. Server-side model methods must use an explicit credential projection so ordinary account-link API responses never expose this ciphertext.

#### 🧪 How to Test

- [x] Tested locally
- [ ] Added/updated tests
- [ ] No tests needed

Migration SQL is idempotent (`IF NOT EXISTS` / `DROP CONSTRAINT IF EXISTS` + `ADD`) and was applied against a dev database via the standard migration flow. `bun run db:generate` reports no further drift.

#### 📝 Additional Information

Migration regenerated on top of `canary` (sequence `0123`) via `bun run db:generate`. Supersedes #17302, which is closed in favor of this consolidated migration.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/database/src/schemas/resourcePermission.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import { timestamps } from './_helpers';
+import { users } from './user';
+import { workspaces } from './workspace';
+
+/**
+ * Entity kinds that support per-resource permission grants. The table is
+ * polymorphic on purpose: adding permission support to a new entity only
+ * requires a new literal here, not a new table.
+ */
+export const PERMISSION_RESOURCE_TYPES = ['agent', 'agentGroup', 'document'] as const;
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/schemas/resourcePermission.ts:12]: Acknowledged — keeping `agentGroup` deliberately. The `chatGroup` literals in `resourceEvents/types.ts` and `editLock` belong to separate subsystems with their own `ResourceType` unions that never read or write `resource_permissions`. All actual consumers of this table (the `lambda/agentGroup` router, conversation resource guard, and the resource-permission model in the follow-up PRs) consistently use `agentGroup`, matching the domain router name already on canary (`apps/server/src/routers/lambda/agentGroup.ts`). Unifying the wider resource-key catalog across resourceEvents/editLock is a broader pre-existing naming split, out of scope for this DB-only PR.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/schemas/resourcePermission.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    id: uuid('id').defaultRandom().primaryKey().notNull(),
+
+    resourceType: text('resource_type', { enum: PERMISSION_RESOURCE_TYPES }).notNull(),
+    resourceId: text('resource_id').notNull(),
+
+    workspaceId: text('workspace_id')
+      .references(() => workspaces.id, { onDelete: 'cascade' })
+      .notNull(),
+
+    accessLevel: text('access_level', { enum: RESOURCE_ACCESS_LEVELS }).notNull(),
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/schemas/resourcePermission.ts:78]: Acknowledged, keeping this at the application layer deliberately. None of the 41 existing schema files in `packages/database/src/schemas` use SQL CHECK constraints — the repo-wide pattern is app-layer validation (Drizzle `text(..., { enum })` is intentionally type-only). All writers go through the resource-permission model, which validates via `isResourceAccessLevelAllowed` exported right here. A CHECK would also break this table's stated polymorphic contract (adding a new resource type = one new literal, no migration): every catalog addition would then require an ALTER CONSTRAINT migration.

**Comment by rdmclin2** [comment]
@codex review latest code

File: packages/database/src/schemas/messengerAccountLink.ts [hunk 3]
```diff
@@ -49,6 +51,21 @@ export const messengerAccountLinks = pgTable(
...
+     */
+    applicationId: varchar('application_id', { length: 255 }),
+
+    /**
+     * AES-GCM encrypted platform credential JSON. WeChat stores
+     * `{ baseUrl, botId, botToken }`; shared-bot platforms leave this null.
+     * Server-side model methods must use an explicit credential projection so
+     * ordinary account-link API responses never expose this ciphertext.
+     */
+    credentials: text('credentials'),
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/schemas/messengerAccountLink.ts:68]: Fixed in eeefb787bc. `MessengerAccountLinkModel` now uses a default safe projection (`getTableColumns` minus `credentials`) for every row-returning `select()`/`returning()`, typed as the new `MessengerAccountLinkPublicItem` (`Omit<MessengerAccountLinkItem, 'credentials'>`). `messenger.getMyLink` / `listMyLinks` and all other model consumers now structurally cannot receive the ciphertext; credential reads will require an explicit credential-scoped method added by the PR that introduces the first writer.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/models/messengerAccountLink.ts [hunk 4]
```diff
@@ -142,7 +148,7 @@ export class MessengerAccountLinkModel {
           workspaceId: params.workspaceId ?? null,
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/models/messengerAccountLink.ts]: Fixed in c8d3f17fb1. Both update branches of `upsertForPlatform` now refresh `applicationId`/`credentials` when the params provide them, and preserve the stored values when the fields are omitted (shared-bot re-verifies pass neither, and the rows we read back are credential-free public projections, so omission must not wipe). Added a regression test covering rotate-and-preserve semantics plus the projection never exposing the ciphertext.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/schemas/messengerAccountLink.ts [hunk 5]
```diff
@@ -49,6 +51,21 @@ export const messengerAccountLinks = pgTable(
     /** Optional platform-side display name (Telegram @username, Slack real_name, etc.) */
     platformUsername: text('platform_username'),
 
+    /**
+     * Platform-side application/bot id for user-scoped credential platforms.
+     * WeChat stores the scanned iLink bot id here so inbound payloads can fail
+     * closed before decrypting credentials. Null for shared-bot platforms.
+     */
+    applicationId: varchar('application_id', { length: 255 }),
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/schemas/messengerAccountLink.ts:60]: Fixed in d6fc2da9f6. Added a partial unique index `messenger_account_links_platform_tenant_application_unique` on `(platform, tenant_id, application_id)` `WHERE application_id IS NOT NULL` — a second link claiming the same credential bot id now fails closed, while shared-bot rows (NULL `application_id`) stay unconstrained. Migration 0123 regenerated to include it, plus a regression test covering both the rejection and the NULL-exempt path.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/schemas/messengerAccountLink.ts [hunk 6]
```diff
@@ -78,6 +96,15 @@ export const messengerAccountLinks = pgTable(
...
     ),
+    // A user-scoped credential bot (WeChat iLink bot id) belongs to exactly
+    // one account link — a second link claiming the same bot id must fail
+    // closed instead of creating an ambiguous inbound route with the wrong
+    // ciphertext. Partial: shared-bot rows keep `application_id` NULL and are
+    // unaffected. Mirrors the unique routing keys on `messenger_installations`
+    // and `agent_bot_providers`.
+    uniqueIndex('messenger_account_links_platform_tenant_application_unique')
+      .on(t.platform, t.tenantId, t.applicationId)
+      .where(isNotNull(t.applicationId)),
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/schemas/messengerAccountLink.ts:107]: Fixed in e5dcf4891b. The upsert catch now has a dedicated branch for `messenger_account_links_platform_tenant_application_unique`: it resolves the claiming link and throws `MessengerAccountLinkConflictError` (with the owning user id) when another user holds the bot, falling back to `MessengerAccountLinkRelinkRequiredError` defensively — so the verify flow gets a typed 409-able conflict instead of the generic final error. Regression test added for the cross-user claim path.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/models/messengerAccountLink.ts [hunk 7]
```diff
@@ -140,9 +174,14 @@ export class MessengerAccountLinkModel {
           platformUsername: params.platformUsername ?? null,
           updatedAt: now,
           workspaceId: params.workspaceId ?? null,
+          // Refresh rotated user-scoped credentials on re-verify; omitting the
+          // fields preserves the stored values (the row shapes we read back
+          // are credential-free public projections, so we can't backfill).
+          ...(params.applicationId === undefined ? {} : { applicationId: params.applicationId }),
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/models/messengerAccountLink.ts]: Fixed in 9cd2c308c4. Extracted the conflict mapping into a shared `throwApplicationClaimConflict` helper (plus a `uniqueViolationConstraint` extractor) and applied it to both credential-refresh update branches, so a rotation onto a bot id claimed by another link now throws the same typed `MessengerAccountLinkConflictError` / `MessengerAccountLinkRelinkRequiredError` as the insert path instead of a raw unique-violation. Regression test added for the rotate-onto-claimed-bot update path.

**Comment by rdmclin2** [comment]
@codex review

File: packages/database/src/models/messengerAccountLink.ts [hunk 8]
```diff
@@ -1,16 +1,31 @@
...
  */
 const GLOBAL_TENANT_ID = '';
 
+const APPLICATION_UNIQUE_CONSTRAINT = 'messenger_account_links_platform_tenant_application_unique';
+
+/** Returns the violated constraint name when `error` is a Postgres unique violation. */
+const uniqueViolationConstraint = (error: unknown): string | undefined => {
+  const pgError = error as { cause?: { code?: string; constraint?: string }; code?: string };
+  const code = pgError.cause?.code ?? pgError.code;
+  return code === '23505' ? pgError.cause?.constraint : undefined;
```
[the remark this answers is not in the document]
Comment (rdmclin2) [reply packages/database/src/models/messengerAccountLink.ts]: Fixed in 034c6904be. `uniqueViolationConstraint` now reads `constraint` from both the top-level error (node-postgres driver) and `cause` (drizzle/pg wrappers), symmetric with how it already read `code`.

**Comment by rdmclin2** [comment]
@codex review
